### PR TITLE
LR Gaussian and test function

### PR DESCRIPTION
**Pull request by CyrusZhang73**

### CODE REVIEW

File: viabel/approximations.py [hunk 1]
```diff
@@ -547,4 +547,127 @@ def _pth_moment(self, var_param, p):
...
 
+    def _kl(self, var_param0, var_param1):
+        param_dict0 = self._pattern.fold(var_param0)
+        param_dict1 = self._pattern.fold(var_param1)
+        mean_diff = param_dict0['mu'] - param_dict1['mu']
+
+        B0 = param_dict0['low_rank']
+        D0 = np.exp(2 * param_dict0['log_sigma'])
+        Sigma0 = np.diag(D0) + np.dot(B0, B0.T)
+        Sigma0_det = np.linalg.det(Sigma0)
```
Comment (jhuggins) [inline viabel/approximations.py:610]: same comment as in the entropy case about efficiently computing the determinant. Also, since this is done in three different places, you should create a separate helper function to do this

File: viabel/approximations.py [hunk 2]
```diff
@@ -547,4 +547,127 @@ def _pth_moment(self, var_param, p):
...
+
+        B1 = param_dict1['low_rank']
+        D1 = np.exp(2 * param_dict1['log_sigma'])
+        D1_ma = np.diag(D1)
+        D1_inv = np.linalg.inv(D1_ma)
```
Comment (jhuggins) [inline viabel/approximations.py:615]: Never compute inverses explicitly. Use `solve` or the Cholesky decomposition to ensure numerical stability

File: viabel/approximations.py [hunk 3]
```diff
@@ -547,4 +547,127 @@ def _pth_moment(self, var_param, p):
...
+        B = param_dict['low_rank']
+
+        return param_dict['mu'] + np.dot(z,B.T) + D_exp * epsilon
+
+    def _entropy(self, var_param):
+        param_dict = self._pattern.fold(var_param)
+        B = param_dict['low_rank']
+        D_exp = np.exp(2 * param_dict['log_sigma'])
+        Sigma = np.diag(D_exp) + np.dot(B, B.T)
+        Sigma_det = np.linalg.det(Sigma)
```
[the remark this answers is not in the document]
Comment (jhuggins) [reply viabel/approximations.py:599]: Also, for numerical stability you should compute the log determinant rather than the determinant

File: viabel/approximations.py [hunk 4]
```diff
@@ -547,4 +547,127 @@ def _pth_moment(self, var_param, p):
...
+
+
+        #KL divergence
+        Sigma_diff = np.log(Sigma1_det / Sigma0_det)
+        Mean_sigma = mean_diff.T @ Sigma_inv @ mean_diff
+        Sigma_trace = np.trace(Sigma_inv @ Sigma0)
+        return .5 * (Sigma_diff - self.dim + Mean_sigma + Sigma_trace)
+
+
+    def log_density(self,var_param, x):
```
Comment (jhuggins) [inline viabel/approximations.py:632]: same comments about determinant and matrix inverse computations apply here as well

Comment (jhuggins) [reply viabel/approximations.py:632]: It might be helpful to look at the scipy multivariate_normal implementation

File: viabel/tests/test_approximations.py [hunk 5]
```diff
@@ -144,3 +144,14 @@ def test_NVP():
             var_param1 = np.random.randn(approx.var_param_dim) / 100
             _test_family(approx, var_param0, var_param1, [])
     # TODO: check behavior in corner cases
+    
+
+def test_LRGaussian():
+    np.random.seed(1214)
+    k=1
```
Comment (jhuggins) [inline viabel/tests/test_approximations.py:151]: it would be good to test for `k > 1` as well

File: viabel/tests/test_approximations.py [hunk 6]
```diff
@@ -144,3 +144,14 @@ def test_NVP():
...
+    for dim in [1, 3]:
```
Comment (jhuggins) [inline viabel/tests/test_approximations.py:152]: and maybe dim = 6 too

File: viabel/tests/test_approximations.py (hunk 5 above)
Comment (jhuggins) [inline viabel/tests/test_approximations.py:151]: use consistent spacing around `=` signs (one space before and after)

**Review by jhuggins — CHANGES_REQUESTED** [review]
This is a good start but need to address a few issues related to numerical stability and computational efficiency

**Comment by CyrusZhang73** [comment]
Thanks for the comprehensive comments. It helps a lot.

File: viabel/approximations.py [hunk 7]
```diff
@@ -555,10 +555,20 @@ def _get_low_rank_mu_sigma_pattern(dim, k):
     ms_pattern['low_rank'] = NumericArrayPattern(shape=(dim, k))
     return ms_pattern
 
-def _get_log_det_sigma(sigma):
-    L = np.linalg.cholesky(sigma)
-    Sigma_det = 2 * np.sum(np.log(np.diag(L)))
-    return Sigma_det
+def matrix_determinant_lemma(D, B):
```
Comment (jhuggins) [inline viabel/approximations.py:558]: You need to document what's actually being computed here

File: viabel/approximations.py [hunk 8]
```diff
@@ -555,10 +555,20 @@ def _get_low_rank_mu_sigma_pattern(dim, k):
...
+    # Compute the determinant of the matrix using the matrix determinant lemma
+    det_D = np.prod(D)
+    inv_D = 1/D
+    det_IpDBBT = np.linalg.det(np.eye(len(D)) + inv_D @ B @ B.T)
+    det_M = det_D * det_IpDBBT
+    return det_M
```
Comment (jhuggins) [inline viabel/approximations.py:564]: You should be computing the log determinant

File: viabel/approximations.py [hunk 9]
```diff
@@ -555,10 +555,20 @@ def _get_low_rank_mu_sigma_pattern(dim, k):
...
+
+def matrix_trace_lemma(D1, B1, D2, B2):
```
Comment (jhuggins) [inline viabel/approximations.py:566]: You need to document what's actually being computed here

File: viabel/approximations.py [hunk 10]
```diff
@@ -662,8 +674,11 @@ def _pth_moment(self, var_param, p):
         param_dict = self._pattern.fold(var_param)
         D_exp = np.exp(2 * param_dict['log_sigma'])
         B = param_dict['low_rank']
-        Sigma = np.diag(D_exp) + np.dot(B, B.T)
-        Scale = np.linalg.eigvalsh(Sigma)
+
+        #eigvals = eigavalsh(B.T @ B) + eigavalsh(D)
+        eig_bbT = np.linalg.eigvalsh(B @ B.T)
+        eig_d = np.sum(D_exp)
+        Scale = eig_bbT + eig_d
```
Comment (jhuggins) [inline viabel/approximations.py:681]: I don't think you can't just add eigenvalues like this as they aren't in the same basis

File: viabel/approximations.py [hunk 11]
```diff
@@ -547,4 +547,142 @@ def _pth_moment(self, var_param, p):
...
+        mean = param_dict['mu']
+        B = param_dict['low_rank']
+        D_exp = np.exp(2*param_dict['log_sigma'])
+        D_inv = np.diag(1/D1)
+
+        Sigma_det = matrix_determinant_lemma(D_exp, B)
+
+        # By the Woodbury formula
+        D_invB = D_inv @ B
+        I_BDB_inv = np.linalg.inv(np.eye(self._k) + np.dot(B.T, D_invB))
```
Comment (jhuggins) [inline viabel/approximations.py:658]: use `solve` not `inv`

File: viabel/approximations.py [hunk 12]
```diff
@@ -547,4 +547,142 @@ def _pth_moment(self, var_param, p):
...
+        Sigma_inv = D_inv - D_invB @ I_BDB_inv @ D_invB.T
```
Comment (jhuggins) [inline viabel/approximations.py:659]: If you are careful about multiplying diff by each term in `Sigma_inv` you can avoid any $O(k^2)$ operations / space requirements

File: viabel/approximations.py [hunk 13]
```diff
@@ -547,4 +547,142 @@ def _pth_moment(self, var_param, p):
...
+
+
+    def log_density(self,var_param, x):
+        if x.ndim == 1:
+            x = x[np.newaxis, :]
+        param_dict = self._pattern.fold(var_param)
+        mean = param_dict['mu']
+        B = param_dict['low_rank']
+        D_exp = np.exp(2*param_dict['log_sigma'])
+        D_inv = np.diag(1/D1)
```
Comment (jhuggins) [inline viabel/approximations.py:652]: You should be able to avoid explicitly instantiating any $k \times k$ matrices

**Review by jhuggins — CHANGES_REQUESTED** [review]
See comments

**Review by jhuggins — CHANGES_REQUESTED** [review]

File: viabel/tests/test_approximations.py [hunk 14]
```diff
@@ -155,3 +155,27 @@ def test_LRGaussian():
             var_param0 = np.random.randn(approx.var_param_dim)
             var_param1 = np.random.randn(approx.var_param_dim)
             _test_family(approx, var_param0, var_param1, [2, 4])
+            
+            
+D = np.array([0.01, 0.02, 0.03]) #log scale
```
Comment (jhuggins) [inline viabel/tests/test_approximations.py:160]: exp(D) is approximately equal to all 1s. Use something like [-1, 0, 1] instead

File: viabel/tests/test_approximations.py [hunk 15]
```diff
@@ -155,3 +155,27 @@ def test_LRGaussian():
...
+B = np.array([[1, 2], [3, 4], [5, 6]])
+D1 = np.array([0.03, 0.02, 0.01]) # log scale
```
Comment (jhuggins) [inline viabel/tests/test_approximations.py:162]: exp(D) is approximately equal to all 1s. Use something like [-1, 0, 1] instead

File: viabel/tests/test_approximations.py [hunk 16]
```diff
@@ -155,3 +155,27 @@ def test_LRGaussian():
...
+B1 = np.array([[6, 5], [4, 3], [2, 1]])
+
+def test_get_log_determinant(D, B):
+    # Expected result calculated manually or using a verified method
+    expected_result = np.log(np.linalg.det(B @ B.T + np.diag(np.exp(2 * D))))
+    actual_result = _get_log_determinant(D, B)
+    p_value = abs(expected_result - actual_result)
```
Comment (jhuggins) [inline viabel/tests/test_approximations.py:169]: this is not a p-value.

File: viabel/tests/test_approximations.py [hunk 17]
```diff
@@ -155,3 +155,27 @@ def test_LRGaussian():
...
+    assert p_value < 0.01
```
Comment (jhuggins) [inline viabel/tests/test_approximations.py:170]: consider using `numpy.testing.assert_allclose`

File: viabel/approximations.py [hunk 18]
```diff
@@ -587,22 +587,20 @@ def _get_trace(D0, B0, D1, B1):
     """
 
     I_B1D1B1 = np.eye(B1.shape[1]) + B1.T / D1 @ B1
-    I_B1D1B1_inv = np.linalg.solve(I_B1D1B1, np.identity(I_B1D1B1.shape[0]))
-
     invD1_B1 = B1 / D1[:, np.newaxis]
-    invD1_B1_I_B1D1B1_inv = invD1_B1 @ I_B1D1B1_inv
+    invD1_B1_I_B1D1B1_inv = (np.linalg.solve(I_B1D1B1.T, invD1_B1.T)).T
```
Comment (jhuggins) [inline viabel/approximations.py:591]: you don't need the outer parentheses here

**Review by jhuggins — CHANGES_REQUESTED** [review]

**Review by jhuggins — APPROVED** [review]
